Implement OpenMP offload for do group update

edoyango opened the pull request:
**Description**
This adds the necessary code to make mpp_do_group_update work with arrays that are managed by NVIDIA's OpenMP offload runtime. This attempts to be minimally disruptive in that non-nvidia compilers will see the same behaviour as previously by adding macros around the relevant openmp directives.

Fixes #1771

**How Has This Been Tested?**
The OpenMP offload capability is currently tested on the "double gyre" case in MOM6-examples using the `nvfortran` compiler and a cuda-aware openmpi. We have [some notes](https://github.com/marshallward/mom6-gpu-report/blob/main/gpu_report.rst) on how to run the gpu-enabled MOM6, but is outdated.

**Checklist:**
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] New check tests, if applicable, are included
- [ ] `make distcheck` passes

edoyango committed:
add gpu2gpu mpi transer with flag for do_group_update

edoyango committed:
add missing collapse(3) clauses

edoyango committed:
Use __NVCOMPILER macro for target regions

edoyango committed:
add back old omp directive wrapped in #ifndef __NVCOMPILER

edoyango committed:
port remaining un/pack loops

edoyango committed:
sub __NVCOMPILER with __NVCOMPILER_OPENMP_GPU

JorgeG94 committed:
add multi gpu support (#2)

* add multi gpu support

* address review comments, add helpful comment for the acc/mp runbtime call

edoyango committed:
allow choice of gpu or cpu parallel

To enable this,  had to be removed -
otherwise segfaults happen on the GPU.

edoyango committed:
fix omp set device call

marshallward commented:
I've finally managed to put together a timing test for MOM6, and the timings for domain decomposition are promising.  There is strong scaling for the larger domains as we saturate the GPU.

<img width="1400" height="800" alt="gpu_mpi_comm" src="https://github.com/user-attachments/assets/ecd68522-5ff1-495a-98ed-409551f09e53" />

Scaling is shown for up to 8 GPUs over 4 nodes on Ursa, so this method should work on our NOAA resources.

Ursa does not have Nvlink, so its comm time shouldn't be taken as reflective of production runs.  But I was surprised that it was not too large, even after going through the PCIe and Infiniband layers.

I will look into including `mpp` tests which exercise this new code.

edoyango committed:
Revert "allow choice of gpu or cpu parallel"

This reverts commit 0cc2a77ee70851ad08458a52a1e4f6307e163463.
Having both the CPU and GPU OpenMP directives compiled caused
a significant slowdown in GPU packing/unpacking performance -
even if parallelism is controlled using OpenMP "if" clause.

marshallward committed:
OMP MPI: Minor cleanups

Some very minor changes to the OpenMP target MPI PR:

* use_device_ptr -> use_device_addr

  This appears to be the updated form (or at least nvfortran says it is)

* Whitespace added to `!$ use omp_lib`

  Does not seem crucial but from our previous discussion it appears more
  correct.

* Removal of some trailing whitespace.

marshallward committed:
OMP target MPI: line length compliance

This patch refactors several lines to keep within the 121-character line
length limit prescribed by the FMS style guidelines.

marshallward committed:
OMP MPI: Update nocomm interface

The no-comm (no MPI) interface has been updated to support the new
omp_offload argument.

bensonr reviewed (commented):
For some of the target omp statements, the `if (use_device_ptr)` is prior to the target data mapping and other times after.  Does it matter where in the !$omp specification the if-test occurs?

The preference for !$omp is `default(none)` with full specification of `shared` and `private`.

edoyango commented on mpp/include/group_update_pack.inc:
```diff
@@ -1,158 +1,233 @@
 
 if( group%k_loop_inside ) then
+#ifndef __NVCOMPILER_OPENMP_GPU
 !$OMP parallel do default(none) shared(npack,group,ptr,nvector,ksize,buffer_start_pos) &
 !$OMP                          private(buffer_pos,pos,m,is,ie,js,je,rotation, &
-!$OMP                                  ptr_field, ptr_fieldx, ptr_fieldy,n,k)
+!$OMP                                  ptr_field, ptr_fieldx, ptr_fieldy,n,k,ni,nj,idx)
+#endif
```
hi @bensonr I've attempted to implement the suggested macro solution, but nvfortran fails to run. 

the fail happens when there's an openmp offload loop and a openmp cpu loop - either with `default(none)` - referencing the same cray pointer in the same procedure. For the latest compilers, whichever loop has `default(none)` either hangs or segfaults at runtime. 

the workaround is to use `default(shared)` in both directives, or to put them into isolated subroutines.

reproducer:
[test_cray_ptr_omp.txt](https://github.com/user-attachments/files/26764707/test_cray_ptr_omp.txt)
compile with `nvfortran -mp=gpu` and run with `./a.out 0/1/2/3`. 0, 1 cases are gpu and cpu (respectively) with `default(none)`, and 2, 3 are with `default(shared)`.

Could a concession be made to use `default(shared)`? otherwise we can't make progress until cray pointers are removed.

bensonr replied:
@edoyango - Thanks for explaining the issue with default(none) and why you'd need default(shared) for now.  It would be helpful to leave the default(none) logic in place along with extra comments for the people in the process of removing cray pointers.

edoyango replied:
@bensonr I've:

* added required shared/private to the openmp offloaded loops
* swapped from `default(none)` to `default(shared)` in the packing/unpacking loops
* added comments to the top of `group_update_un/pack.inc` files explaining the restrictions imposed by nvfortran with cray pointers
* through macro manipulation, ensured openmp cpu parallelism is activated when `ompoffload==.false.`

I hope that's sufficient.

bensonr commented on mpp/include/mpp_transmit.inc:
```diff
@@ -170,7 +171,7 @@
     end subroutine MPP_SEND_
 
-    subroutine MPP_RECV_SCALAR_( get_data, from_pe, glen, block, tag, request )
+    subroutine MPP_RECV_SCALAR_( get_data, from_pe, glen, block, tag, request, omp_offload )
```
@edoyango - I see you only added the offload option for the scalar version of the send/recv pairs.  Is there a reason this is not being done for all of the send/recv routines?

edoyango replied:
I've tried to keep the scope of this PR to only cover what's required to get `do_group_update` working and `do_group_update` only uses `MPP_SEND/RECV_SCALAR_`. I can add the `omp_offload` flag to the other implementations, but it would be untested.

bensonr replied:
Thanks for the update.  I'll discuss with the rest of the team how we want to handle the other routines.

Now that you've mentioned testing, will you be adding or adapting any of the existing unit tests to add GPU-specific tests?

edoyango replied:
|> Now that you've mentioned testing, will you be adding or adapting any of the existing unit tests to add GPU-specific tests?

Sorry for the late reply! I just pushed some tests to this branch. It co-opts the existing mpp domain tests by adding a test_group_offload switch (thanks @rem1776 for mocking one together). The test passes for both nvhpc 26.3. gcc passes as well, although it doesn't actually offload to GPU. 

Note that the domain tests include non-blocking tests, but the openmp offload version of those haven't been implemented, so those tests just run on CPU.

edoyango committed:
use openmp cpu if ompoffload=.false.

This ensures that (un)packing steps in do_group_update is performed
with openmp cpu parallelism if ompoffload=.false.. Previously it
would only do serial. This is implemented by undefining the GPU
macro (currently __NVCOMPILER_OPENMP_GPU) and re-including the
(un)packing files.

To make this work, the default(shared) was used in all the relevant
OpenMP directives. If default(none) is used, the loops would hang
or segfault.

Commit:
add changes for group update test

edoyango committed:
remove trailing whitespace in comments

edoyango committed:
fix k outside path and normalise omp offload clauses

edoyango committed:
offload folded-north post-processing

Did't use the if(use_device_ptr) omp clause because
the tests would hang. Instead made explicit branches.

Left out mappings because gcc didn't like cray pointers
in maps

edoyango committed:
enable group update omp offload tests

Update test_mpp_domains to pass the OpenMP offload flag through the group
update path and stage test data on the device for offloaded runs.

Add explicit OpenMP maps and shared clauses for the offloaded setup and
reference-copy loops, including shifted CGRID and BGRID extents.

Unskip the group update OpenMP offload test in test_mpp_domains2.sh.

Commit:
add select-rank-less versions of mpp_pack and mpp_global_field

Commit:
quick fix for class(*) in diag_data

Commit:
remove whitespace

Commit:
fix indent in diag_data

Commit:
add error message and move ifdef

edoyango committed:
mpp: fix uninitialized error text on single-PE runs

Commit 22dfe08 redirected the mpp_error_basic label into a new
text_errortype variable but only copied it into the printed `text`
inside the npes>1 branch. On single-PE runs `text` was left
uninitialized, so mpp_error emitted garbled stack memory. Seed `text`
from text_errortype on the npes==1 path.

uramirez8707 reviewed (dismissed):
Thank you @edoyango 

Looks good to me.
